File: danielle_yahtzee_cpp_code/ish_df_yahtzee_v4.6/ScoreCard.cpp
#include <iostream>
#include <iomanip>
#include <algorithm>
#include <vector>
#include <numeric>
#include "ScoreCard.h"
#include "Dice.h"

using namespace std;


//************************************************************
//                  Default Constructor    
//************************************************************

ScoreCard::ScoreCard() {
    
    for (int i = 0; i < 13; i++) {
        scores[i] = -1;
        isSelected[i] = 0;
    }
   //printScoreCard();
}


//************************************************************
//              Set points in scores array           
//************************************************************

void ScoreCard::recordScore(int category, int score) {
    
    static int round = 0;
    
    //cout<<"\nInside recordScore()  "; get_isSelected();   
    //cout << "round = " << round << endl;
    
    if(round >= 0 && round <= 12){ scores[category-1] = score; }
  
    else if( round > 12 && (isSelected[category-1] == 0) ) {
        scores[category-1] = score;
    }
    else { scores[category-1] = -1;}
   
    round++;
}


//************************************************************
//          Add upper and lower sections of scorecard            
//************************************************************

int ScoreCard::getTotalScore() {
    
    int total = 0;    
    for (int i = 0; i < 13; i++) {        
        if (scores[i] >= 0) {
            
            // if upper section is 63 or more than add 25 pts to total
            if(i==5 && total >= 63){ total +=25; } 
            total += scores[i];
        }
    }
    return total;
}

//*****************************************************************************
//        Set player's scorecard using values from possible scorecard
//*****************************************************************************

void ScoreCard::setFinalSC(ScoreCard &scorecard, int diceArr[]){
    
    int choice = 0;
     
    cout << "\nPick a Category between 1 and 13.\n";
    
    do {
       
        cin >> choice;

        // Check if the choice is valid and not already used.
        // You need to implement the logic to store the used categories and check for valid scoring.
        while (choice < 1 || choice > 13) {
            cout << "Invalid choice. Please select a category between 1 and 13." << endl;
            cin >> choice;
        }

        // Conditional stops a set category from being overwritten
        if( (isSelected[choice-1]) != 0) { 
            cout << "Category " << choice << " has already been selected.\n"
                 << "Chose another category:  ";
        }
    } while( (isSelected[choice-1]) != 0);
    
    // Set points in player's final scoreboard
    scores[choice-1] = scorecard.scores[choice-1];     
            
    // set flag so this category, so it can't be selected anymore
    isSelected[choice-1] = true;
    scorecard.isSelected[choice-1] = true;
    
    //cout<<"\nInside setFinalSC()  "; get_isSelected();    
    //getFinalSC();
}


//************************************************************
//               Fill scorecard with possible points
//************************************************************

void ScoreCard::fillScoreCard(int diceArr[], bool clearance) {
    
    int num_dice = 5;
    vector<int> scores(13, 0);
    vector<int> counts(6, 0);    
    
    // Sort array in ascending order
    sort(diceArr, diceArr + num_dice);

    // ? Accumulating something in scores[]
    for (int i = 0; i < num_dice; i++) {
        counts[diceArr[i] - 1]++;
        scores[diceArr[i] - 1] += diceArr[i];
    }

    // ? Record scores to upper section?  
    for (int i = 0; i < 6; i++) {
        recordScore(i + 1, scores[i]);
    }

    //  Check Lower section of scorecard  starts here  
    bool three_of_a_kind = false;
    bool four_of_a_kind = false;
    bool full_house = false;
    bool small_straight = false;
    bool large_straight = false;
    bool isYahtzee = true;

    // Check for Three and four of a kind
    for (int count : counts) {
        if (count >= 3) {
            three_of_a_kind = true;
            if (count >= 4) {
                four_of_a_kind = true;
                break;
            }
        }
    }

    // if 3 or 4 of kind is true, then set pts in scores[]
    recordScore(7, three_of_a_kind ? accumulate(diceArr, diceArr + num_dice, 0) : 0);
    recordScore(8, four_of_a_kind ? accumulate(diceArr, diceArr + num_dice, 0) : 0);

    // if Full house, then set pts in scores[]
    full_house = (counts[0] == 3 && counts[1] == 2) || (counts[0] == 2 && counts[1] == 3);
    recordScore(9, full_house ? 25 : 0);

    // Check for Small and large straight
    for (int i = 0; i < 2; i++) {
        if (counts[i] && counts[i + 1] && counts[i + 2] && counts[i + 3]) {
            small_straight = true;
            if (counts[i + 4]) {
                large_straight = true;
                break;
            }
        }
    }

    // if small or large straight, then set pts in scores[]
    recordScore(10, small_straight ? 30 : 0);
    recordScore(11, large_straight ? 40 : 0);

    // if Yahtzee, then set pts in scores[]
    isYahtzee = all_of(counts.begin(), counts.end(), [&](int count) { return count == num_dice; });
    recordScore(12, isYahtzee ? 50 : 0);

    // Chance set pts in scores[]
    recordScore(13, accumulate(diceArr, diceArr + num_dice, 0));    

    //cout<<"\nPoints Possible "; printScoreCard();     
}


//************************************************************
//               Print bool isSelected array
//************************************************************

void ScoreCard::get_isSelected() const {
    cout<<"\n isSelected[]";
    for(int i=0; i<13;i++){
        cout << isSelected[i] << " ";
    }
}


//************************************************************
//              Print scorecard and total points           
//************************************************************

void ScoreCard::getFinalSC(){
    cout << "\n\t\tScorecard\n";
    printScoreCard();
    cout << setw(31) << "Total Score: " << setw(3) << getTotalScore() << "\n";
}



//************************************************************
//                  Print scorecard        
//************************************************************

void ScoreCard::printScoreCard() const {
    cout << setw(5) << " "  << "-------------------------------\n";
    cout << setw(5) << " "  << "| Category          | Score   |\n";
    cout << setw(5) << " "  << "-------------------------------\n";
    cout << setw(5) << " "  << left << setw(20) << "| 1. Ones" << " | " << right << setw(6) << (scores[0] >= 0 ? to_string(scores[0]) : "-") << " |\n";
    cout << setw(5) << " "  << left << setw(20) << "| 2. Twos" << " | " << right << setw(6) << (scores[1] >= 0 ? to_string(scores[1]) : "-") << " |\n";
    cout << setw(5) << " "  << left << setw(20) << "| 3. Threes" << " | " << right << setw(6) << (scores[2] >= 0 ? to_string(scores[2]) : "-") << " |\n";
    cout << setw(5) << " "  << left << setw(20) << "| 4. Fours" << " | " << right << setw(6) << (scores[3] >= 0 ? to_string(scores[3]) : "-") << " |\n";
    cout << setw(5) << " "  << left << setw(20) << "| 5. Fives" << " | " << right << setw(6) << (scores[4] >= 0 ? to_string(scores[4]) : "-") << " |\n";
    cout << setw(5) << " "  << left << setw(20) << "| 6. Sixes" << " | " << right << setw(6) << (scores[5] >= 0 ? to_string(scores[5]) : "-") << " |\n";
    cout << setw(5) << " "  << "-------------------------------\n";
    cout << setw(5) << " "  << left << setw(20) << "| 7. Three of a Kind" << " | " << right << setw(6) << (scores[6] >= 0 ? to_string(scores[6]) : "-") << " |\n";
    cout << setw(5) << " "  << left << setw(20) << "| 8. Four of a Kind" << " | " << right << setw(6) << (scores[7] >= 0 ? to_string(scores[7]) : "-") << " |\n";
    cout << setw(5) << " "  << left << setw(20) << "| 9. Full House" << " | " << right << setw(6) << (scores[8] >= 0 ? to_string(scores[8]) : "-") << " |\n";
    cout << setw(5) << " "  << left << setw(20) << "| 10. Small Straight" << " | " << right << setw(6) << (scores[9] >= 0 ? to_string(scores[9]) : "-") << " |\n";
    cout << setw(5) << " "  << left << setw(20) << "| 11. Large Straight" << " | " << right << setw(6) << (scores[10] >= 0 ? to_string(scores[10]) : "-") << " |\n";
    cout << setw(5) << " "  << left << setw(20) << "| 12. Yahtzee" << " | " << right << setw(6) << (scores[11] >= 0 ? to_string(scores[11]) : "-") << " |\n";
    cout << setw(5) << " "  << left << setw(20) << "| 13. Chance" << " | " << right << setw(6) << (scores[12] >= 0 ? to_string(scores[12]) : "-") << " |\n";
    cout << setw(5) << " "  << "-------------------------------\n";
}

File: danielle_yahtzee_htmlCode/yahtzee_code_html/yahtzee_html_v7_noPrompts/public_html/assets/javascript/ScoreCard.js
// Global constants for ScoreCard Class
var NUM_DICE = 5;
var NUM_CATGRY = 13; // number of scoring categories

// ScoreCard object constructor. Public properties of ScoreCard class
function ScoreCard() {
    
    //console.log("Hit ScoreCard() Constructor");
    this.setFName("");
    this.round=0; //this.setRound(0);
    this.upperScore = 0;
    this.lowerScore = 0;
    this.isCatgryPicked = [NUM_CATGRY]; // Stops categories from being overwritten or displayed twice
    this.scores = [NUM_CATGRY];     // points for each of 13 categories in scorecard
    this.dice = [NUM_DICE]; // ? Are we using this cuz we have a local one in play() or do we need to connect it with local in play()
    this.selected = [];  // Create vector for the this.dice we're keeping for final points

    // Initialize array of Dice class object
    for (var i = 0; i < NUM_DICE; i++) {
        this.dice[i] = 0;
        this.printDice(i); 
        //console.log("ScoreCard() dice["+i+"]="+this.dice[i]);
    }
    
    // Initialize arrays
    for (var c = 0; c < NUM_CATGRY; c++) {
        this.scores[c] = -1;
        this.writeAScore(c); // Writes element to ScoreCard.html
        this.isCatgryPicked[c] = false;
    }
};



//************************************************************
//      Write one scores element to ScoreCard.html
//************************************************************
ScoreCard.prototype.writeAScore = function (indx) {    
    //console.log("Hit writeAScore()");    
    document.getElementById('catgy'+indx).innerHTML = this.scores[indx]; // display the value on the button   
};



//************************************************************
//               Fill scorecard with possible points
//************************************************************
ScoreCard.prototype.fillScoreCard = function (name) {    
    
    this.setFName(name);
    var NUM_DICE_SIDES = 6;
    var temp_scores = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];   
    var counts = [0, 0, 0, 0, 0, 0];
    this.dice.sort(); // Sort array in ascending order  

    // Loops 5 times and is Accumulating how many dice are the same and their total pts
    for (var i = 0; i < NUM_DICE; i++) {
        var index = this.dice[i] - 1;
        counts[index]++;                     
        temp_scores[index] += this.dice[i]; 
        //console.log("index = " + index + "  this.dice[" + i + "] = " + this.dice[i]);
    }
    
    // Set & write possible points to scorecard's upper section
    for (var i = 0; i < NUM_DICE_SIDES; i++) {
        this.recordScore(i + 1, temp_scores[i]);
    }

    //  Set flags for checking Lower section of scorecard starts here 
    var three_of_a_kind = false;
    var four_of_a_kind = false;
    var full_house = false;
    var small_straight = false;
    var large_straight = false;
    var yahtzee = false;
    var chance = 0;
     
    // Check for Three and four of a kind
    for (var i = 0; i < NUM_DICE_SIDES; i++) {
        if (counts[i] >= 3) {
            three_of_a_kind = true;
             //console.log("three_of_a_kind = true");
            if (counts[i] >= 4) {
                four_of_a_kind = true;
                //console.log("four_of_a_kind = true");
                if (counts[i] === NUM_DICE) {
                    yahtzee = true;
                }
            }
        }
        chance += i+1 * counts[i]; // Each die contributes its face value multiplied by its count to the chance score
    }

    // 3 or 4 of kind is true, then set pts in scores[] 
    this.recordScore(7, three_of_a_kind ? chance : 0); // Assuming score categories are 1-indexed
    this.recordScore(8, four_of_a_kind ? chance : 0);
    
    // Full house, then set pts in scores[]
    if ((counts.includes(2) && counts.includes(3)) || yahtzee) {
        full_house = true;
    }
    this.recordScore(9, full_house ? 25 : 0);
    
    // Small and Large straight
    for (var i = 0; i < 3; i++) {
        if (counts[i] && counts[i+1] && counts[i+2] && counts[i+3]) {
            small_straight = true;
            if (counts[i+4]) {
                large_straight = true;
                break;
            }
        }
    }
    
    // Small or Large straight, then set pts in scores[]
    this.recordScore(10, small_straight ? 30 : 0);
    this.recordScore(11, large_straight ? 40 : 0);
    
    // Yahtzee, then set pts in scores[]
    this.recordScore(12, yahtzee ? 50 : 0);
    
    // Chance set pts in scores[]
    this.recordScore(13, chance);

   console.log("Hit fillScorecard(). Points Possible...[ "+ this.scores+" ]");   
};


//************************************************************
//     Set scores in the possible points scorecard        
//************************************************************

ScoreCard.prototype.recordScore = function (category, score) {
    //console.log("Inside recordScore().  scorecard.round = " + this.getRound());
    
    var string1 = this.fName + "'s Possible ScoreCard ";
    document.getElementById('finalSCDiv').innerHTML = string1;  
    
    var indx = category - 1; //ScoreCard's range [0,12]
    
    // if it's not last round, then set the scores array with possible points
    if (this.getRound() >= 0 && this.getRound() <= 12) {
        this.scores[indx] = score;
        this.writeAScore(indx); // rewrite points on ScoreCard.html
    }

    // if it is last round AND this category has NOT be selected, then set scores array with possible points
    else if (this.getRound() > 12 && (this.isCatgryPicked[category - 1] === false)) {
        this.scores[indx] = score; // on the last round, it will set the unpicked category to zero instead of -1. Every category has to be picked.
        this.writeAScore(indx); // rewrite points on ScoreCard.html
        
    } else { // -1 sets this category to null value, so they can't choose it from possible scorecard
        this.scores[indx] = -1;
        this.writeAScore(indx); // rewrite points on ScoreCard.html
    } 
};



//*****************************************************************
//             Keep this dice in selected vector
//*****************************************************************
ScoreCard.prototype.pushThisDice = function (choice) {
    //choice -= 1;
    this.selected.push(choice);     
    console.log("      Hit  ScoreCard.pushThisDice().  this.selected = dice" + this.selected);
};



//*****************************************************************
//               Set Dice class array
//*****************************************************************
ScoreCard.prototype.reRoll = function () {

    console.log("Hit reRoll().  this.selected dice = " + this.selected);

    // Re-Roll Dice. loops through selected vector, and checks which this.dice # to re-roll ?
    for (var j = 0; j < NUM_DICE; j++) {

        // https://www.w3schools.com/jsref/jsref_find.asp
        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find
        // syntax: array.find(function(currentValue, index, arr),thisValue)
        //var n = this.selected.find(this.isFound(j), j);
        //console.log("Hit reRoll()...keeping dice" + (n));
        
        
        //if(this.selected.find(this.selected.begin(), this.selected.end(), j) === this.selected.end()) {

            //console.log("...Re-Rolling Dice"+j);
            this.roll(j); // re-roll this this.dice
        //}
    }
    this.dice.sort();
    this.printDice();      // Write all dice to html
    this.consoleLogDice();
};


//************************************************************
//           Called in reRoll()
//************************************************************
ScoreCard.prototype.isFound = function (indx) {
    console.log("   Hit isFound(). dice " + indx + " is in this.isCatgryPicked[ ]");
    //var i = (indx === isCatgryPicked[indx]) ? indx : undefined;
    return indx >= 0;
};


//************************************************************
//              Roll & set a single dice
//************************************************************
ScoreCard.prototype.roll = function (i) {
    this.dice[i] = (1 + (Math.floor(Math.random() * 6)));
};



//************************************************************
//              Write 1 dice's value to html
//************************************************************
ScoreCard.prototype.writeADice = function (i) {
    //console.log("Hit writeADice()");  
    var diceBtn = document.getElementById("dice"+i);
    var val = this.getDice(i);
    diceBtn.value = val;
    diceBtn.innerHTML = val;
    //console.log("Hit writeADice().   dice["+i+"] = "+val);
};



//**********************************************************
//                  Write all dice to html
//**********************************************************
ScoreCard.prototype.printDice = function () {
    //console.log("Hit printDice()");
    for (var i = 0; i < NUM_DICE; i++) {
        this.writeADice(i);
        //console.log("Dice[" +i+ "]="+this.dice[i]);
    }
};



//**********************************************************
//          console log all dice 
//**********************************************************
ScoreCard.prototype.consoleLogDice = function () {
    console.log("Hit consoleLogDice().   Rerolled Dice[ "+this.dice+" ]");
};



//************************************************************
//          Calculate totals for upper, lower scorecard sections            
//************************************************************

ScoreCard.prototype.setUpLowSums = function () {

    //console.log("Hit setUpLowSums()");
    // Calculate upper section total points
    var total = 0;
    for (var i = 0; i < 6; i++) {

        if (this.scores[i] >= 0) { // only add values over zero
            total += this.scores[i];
        }

        if (i === 5) { // last points category in upper section of scorecard
            this.upperScore = total;
            //alert("Hit i==5. upperScore=" + this.upperScore +" total="+total);
        }
    }

    // Calculate lower section total points
    total = 0;
    for (var i = 6; i < NUM_CATGRY; i++) {

        if (this.scores[i] >= 0) { // only add values over zero
            total += this.scores[i];
        }
        if (i === 12) { // last points category in lower section of scorecard
            this.lowerScore = total;
            //alert("Hit i==12. lowerScore="+ this.lowerScore +" total="+total);
        }
    }
    //console.log("Hit setUpLowSums().  upperScore = " + this.upperScore + "  lowerScore = " + this.lowerScore);
};


//************************************************************
//          Add upper and lower sections of scorecard  
//      Check if upper section qualifies for bonus points          
//************************************************************

ScoreCard.prototype.getTotalScore = function () {

    var total = 0, bonus=25;

    // if upper section is 63 or more than add 25 pts to total
    if (this.upperScore >= 63) {
        total += bonus;
    }
    total = this.upperScore + this.lowerScore;
    return total;
};


//**********************************************************
//      Prints this player's final ScoreCard Class members
//**********************************************************
ScoreCard.prototype.printFinalSC = function (name) {

    var bonusMin = 63;
    var bonus = (this.getUpperScore() >= bonusMin) ? 25 : 0;
    var string1 = name + "'s Final ScoreCard ";
    this.writeScoreCard ();
    this.setUpLowSums();
    document.getElementById('finalSCDiv').innerHTML = string1;    
    document.getElementById('catgy13').innerHTML = this.getUpperScore();
    document.getElementById('catgy14').innerHTML = bonus;
    document.getElementById('catgy15').innerHTML = this.getLowerScore();
    document.getElementById('catgy16').innerHTML = this.getTotalScore();   
    console.log("Hit printFinalSC(). "+string1+" [ "+this.scores+" ]\n"  
                + "upperScore="+this.getUpperScore() +"  lowerScore="+this.getLowerScore()+" total="+this.getTotalScore());
};


//************************************************************
// Print scores array. Each element represents 1 points category on ScoreCard
//************************************************************
ScoreCard.prototype.writeScoreCard = function () {
    //console.log("Hit writeScoreCard(). scores[ "+ this.scores+" ]"); 
    for (var i = 0; i < NUM_CATGRY; i++) {
        this.writeAScore(i); // Writes element to ScoreCard.html
    }
};



//************************************************************
//              Set bool isCatgryPicked[13]
//************************************************************
ScoreCard.prototype.setWasPicked = function (i) {
    this.isCatgryPicked[i] = true;
    console.log("isCatgryPicked[" + i + "] = "+ this.isCatgryPicked[i]);
};



//************************************************************
//                  Set which round it is
//************************************************************
ScoreCard.prototype.setRound = function (r) {
    this.round = r;   
    //console.log("Hit setRound(). round = " + this.getRound());
    this.writeRound();
};

ScoreCard.prototype.setFName = function (name) {
    this.fName=name;
};

ScoreCard.prototype.writeRound = function() {
    //console.log("Hit writeRound(). round = " + this.getRound());
    var string = "Round " + this.getRound();
    document.getElementById("roundDiv").innerHTML = string;//.append(string);
};

//**********************************************************
//              Return one element in scores array
//**********************************************************
ScoreCard.prototype.getScoreIndex = function (i) {
    console.log("Hit getScoreIndex(). score[] = "+ this.scores[i]);
    return this.scores[i];
};

ScoreCard.prototype.getDice = function (i) {    
    return this.dice[i];
};


//************************************************************
//              Return round
//************************************************************
ScoreCard.prototype.getRound = function () {
    return this.round;
};


//************************************************************
//                  Return upperScore
//************************************************************
ScoreCard.prototype.getUpperScore = function () {
    return this.upperScore;
};

//************************************************************
//                  Return upperScore
//************************************************************
ScoreCard.prototype.getLowerScore = function () {
    return this.lowerScore;
};